File: __tests__/getAmountOut.cpp
#define CATCH_CONFIG_MAIN

#include "catch.hpp"

int64_t getAmountOut( const int64_t amountIn, const int64_t reserveIn, const int64_t reserveOut ) {
    const int64_t amountInWithFee = amountIn * 997;
    const int64_t numerator = amountInWithFee * reserveOut;
    const int64_t denominator = reserveIn * 1000 + amountInWithFee;
    const int64_t amountOut = numerator / denominator;
    return amountOut;
}

TEST_CASE( "getAmountOut (pass)" ) {
    // Inputs
    const int64_t amountIn = 10000;
    const int64_t reserveIn = 45851931234;
    const int64_t reserveOut = 125682033533;

    // Calculation
    const int64_t amountOut = getAmountOut( amountIn, reserveIn, reserveOut );

    REQUIRE( amountOut == 27328 );
}

File: __examples__/get_amount_out.js

// calculations
const fee = 30;
const amount_in = 3734534447974;
const reserve_in = 33593153629677144;
const reserve_out = 899196436;

const amount_in_with_fee = amount_in * (10000 - fee);
const numerator = amount_in_with_fee * reserve_out;
const denominator = (reserve_in * 10000) + amount_in_with_fee;
const amount_out = numerator / denominator;
console.log(amount_out);
// => 99652.29296384696

// 373786282494
// 373786282495